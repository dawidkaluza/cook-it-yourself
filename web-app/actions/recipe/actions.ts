"use server"

type Recipes = {
  items: [
    id: number,
    name: string,
    description: string
  ]
};

export const getMyRecipes = async () : Promise<Recipes> => {

};
