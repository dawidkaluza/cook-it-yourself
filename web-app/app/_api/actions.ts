"use server"

import {cookies} from "next/headers";

export function isSignedIn(){
  const cookieStore = cookies();
  const sessionCookie = process.env.NEXT_PUBLIC_API_GATEWAY_SESSION_COOKIE;
  return sessionCookie ? cookieStore.has(sessionCookie) : false;
}