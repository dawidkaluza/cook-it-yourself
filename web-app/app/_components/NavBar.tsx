"use client"

import React from "react";
import Link from "next/link";
import {usePathname} from "next/navigation";
import {useAuth} from "@/app/_hooks/auth/useAuth";

const NavBar = () => {
  const { isSignedIn } = useAuth();

  return (
    <nav className="navbar navbar-expand bg-body-tertiary">
      <div className="container-fluid text-center">
        <Link href="/" className="navbar-brand">Cook it yourself</Link>
        {isSignedIn ? <NavBarMenu /> : <SignInButton />}
      </div>
    </nav>
  );
};

const NavBarMenu = () => {
  return (
    <>
      <button className="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar-menu"
              aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation">
        <span className="navbar-toggler-icon"></span>
      </button>

      <div className="collapse navbar-collapse" id="navbar-menu">
        <ul className="navbar-nav">
          <NavBarItem path="/my-recipes">My recipes</NavBarItem>
          <NavBarItem path="/cooking">Cooking</NavBarItem>
        </ul>
      </div>
    </>
  );
};

const NavBarItem = ({
  path,
  children
}: Readonly<{
  path: string,
  children: React.ReactNode,
}>) => {
  const currentPath = usePathname();

  return (
    <li className="nav-item">
      <Link className={path === currentPath ? "nav-link active" : "nav-link"} href={path}>{children}</Link>
    </li>
  );
};

const SignInButton = () => {
  return (
    <a
      href={process.env.API_GATEWAY_CLIENT_URL + "/oauth2/authorization/ciy"}
      className="btn btn-primary btn-sm"
    >
      Sign in
    </a>
  );
};

export {NavBar};