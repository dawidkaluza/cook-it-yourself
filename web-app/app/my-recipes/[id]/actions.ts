"use server";

import {RecipeDetails} from "@/app/my-recipes/_dtos/recipe";
import { fetchFromServer} from "@/app/_api/fetch";

export async function reviewRecipe(id: number) : Promise<RecipeDetails> {
  return await fetchFromServer<RecipeDetails>({
    endpoint: `/kitchen/recipe/${id}`
  }) as RecipeDetails;
}