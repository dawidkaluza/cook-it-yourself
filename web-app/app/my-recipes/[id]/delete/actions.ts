"use server";

import {ApiError, fetchFromServer} from "@/app/_api/fetch";
import {redirect} from "next/navigation";

export async function deleteRecipe(prevState: any, formData: FormData): Promise<string> {
  const id = formData.get("id");

  try {
    await fetchFromServer({
      endpoint: `/kitchen/recipe/${id}`,
      method: "DELETE"
    });
  } catch (error) {
    if (error instanceof ApiError) {
      const response = error.response;
      switch (response.status) {
        case 404: {
          return "The recipe is already deleted.";
        }

        case 422: {
          return "Couldn't process the request.";
        }
      }
    }

    throw error;
  }

  redirect("/my-recipes");
}