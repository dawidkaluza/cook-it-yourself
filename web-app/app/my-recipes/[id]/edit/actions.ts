"use server";

import {FieldError} from "@/app/my-recipes/_dtos/errors";
import {ApiError, fetchFromServer} from "@/app/_api/fetch";
import {PortionSize, RecipeDetails, UpdateRecipeRequest} from "@/app/my-recipes/_dtos/recipe";
import {redirect} from "next/navigation";

async function editRecipe(prevState: any, formData: FormData): Promise<FieldError[]> {
  const id = formData.get("id") as string;

  const mapToPortionSize = (formData: FormData): PortionSize => {
    const portionSizeInput = formData?.get("portionSize")?.toString();
    if (!portionSizeInput) {
      return {
        value: "",
        measure: "",
      };
    }

    const numberIndex = portionSizeInput.search(/\d/g);
    if (numberIndex === -1) {
      return {
        value: "",
        measure: "",
      };
    }

    const nonNumberIndex = portionSizeInput.search(/\D/g);
    if (nonNumberIndex === -1 || nonNumberIndex < numberIndex) {
      return {
        value: portionSizeInput.slice(numberIndex),
        measure: "",
      };
    }

    return {
      value: portionSizeInput.slice(numberIndex, nonNumberIndex).trim(),
      measure: portionSizeInput.slice(nonNumberIndex).trim()
    };
  }

  const requestData: UpdateRecipeRequest = {
    basicInformation: {
      name: formData.get("name") as string,
      description: formData.get("description") as string,
      cookingTime: Number(formData.get("cookingTime")),
      portionSize: mapToPortionSize(formData),
    },
    ingredients: {
      ingredientsToAdd: [],
      ingredientsToUpdate: [],
      ingredientsToDelete: [],
    },
    steps: {
      stepsToAdd: [],
      stepsToUpdate: [],
      stepsToDelete: [],
    }
  };

  try {
    const updatedRecipe = await fetchFromServer<RecipeDetails>({
      endpoint: `/kitchen/recipe/${id}`,
      method: "PUT",
      body: requestData
    });

    if (updatedRecipe) {
      redirect(`/my-recipes/${updatedRecipe.id}/`);
    }
  } catch (error) {
    if (error instanceof ApiError) {
      const response = error.response;

      // TODO cover other possible client errors
      if (response.status === 422) {
        const body = await response.json();
        const fields : FieldError[] = body.fields;
        return fields.map((field) => {
          const name = field.name;
          const message = field.message;
          const indexOfDot = name.indexOf(".");
          if (indexOfDot === -1) {
            return {
              name,
              message
            }
          } else {
            const mainName = name.slice(0, indexOfDot);
            return {
              name: mainName,
              message
            };
          }
        });
      }

      return [];
    }

    throw error;
  }

  throw new Error("Couldn't process the request");
}

export { editRecipe };