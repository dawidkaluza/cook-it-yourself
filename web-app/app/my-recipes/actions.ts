"use server";

import {fetchFromServer} from "@/app/_api/fetch";
import {Page} from "@/app/my-recipes/_dtos/page";
import {Recipe} from "@/app/my-recipes/_dtos/recipe";

export async function getMyRecipes(){
  return await fetchFromServer({ endpoint: "/kitchen/recipe" }) as Promise<Page<Recipe>>;
}