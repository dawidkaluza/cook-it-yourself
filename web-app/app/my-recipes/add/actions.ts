"use server";

import {Ingredient, PortionSize, Recipe, Step} from "@/app/my-recipes/add/_dtos/Recipe";
import {ApiError, fetchFromServer} from "@/app/_api/fetch";
import {redirect} from "next/navigation";

export type FieldError = {
  name: string;
  message: string;
};

export async function addRecipe(prevState: any, formData: FormData) : Promise<FieldError[]> {
  const mapToIngredients = (formData: FormData) : Ingredient[] => {
    const ingredientsNames  = formData.getAll("ingredientName") as string[];
    const ingredientsAmounts = formData.getAll("ingredientAmount") as string[];
    const ingredientsUnits = formData.getAll("ingredientUnit") as string[];
    const ingredients : Ingredient[] = [];
    const ingredientsNum = ingredientsNames.length;
    for (let i = 0; i < ingredientsNum; i++) {
      const ingredient = {
        name: ingredientsNames[i],
        value: ingredientsAmounts[i],
        measure: ingredientsUnits[i],
      };
      ingredients.push(ingredient);
    }
    return ingredients;
  };

  const mapToSteps = (formData: FormData) : Step[] => {
    return formData.getAll("methodSteps").map(methodStep => {
      return {
        text: methodStep as string,
      };
    });
  };

  const mapToPortionSize = (formData: FormData) : PortionSize => {
    const portionSizeInput = formData?.get("portionSize")?.toString();
    if (!portionSizeInput) {
      return {
        value: "",
        measure: "",
      };
    }

    const numberIndex = portionSizeInput.search(/\d/g);
    if (numberIndex === -1) {
      return {
        value: "",
        measure: "",
      };
    }

    const nonNumberIndex = portionSizeInput.search(/\D/g);
    if (nonNumberIndex === -1 || nonNumberIndex < numberIndex) {
      return {
        value: portionSizeInput.slice(numberIndex),
        measure: "",
      };
    }

    return {
      value: portionSizeInput.slice(numberIndex, nonNumberIndex).trim(),
      measure: portionSizeInput.slice(nonNumberIndex).trim()
    };
  }

  let recipe : Recipe = {
    name: formData.get("name") as string,
    description: formData.get("description") as string,
    ingredients: mapToIngredients(formData),
    methodSteps: mapToSteps(formData),
    cookingTime: Number(formData.get("cookingTime")),
    portionSize: mapToPortionSize(formData)
  };

  try {
    const createdRecipe = await fetchFromServer<Recipe>({
      endpoint: "/kitchen/recipe",
      method: "POST",
      body: recipe,
    });
    if (createdRecipe) {
      redirect(`/my-recipes/${recipe.id}/`);
    }
  } catch (error) {
    if (error instanceof ApiError) {
      const response = error.response;

      // TODO cover other possible client errors
      if (response.status === 422) {
        const body = await response.json();
        const fields : FieldError[] = body.fields;
        return fields.map((field) => {
          const name = field.name;
          const message = field.message;
          const indexOfDot = name.indexOf(".");
          if (indexOfDot === -1) {
            return {
              name,
              message
            }
          } else {
            const mainName = name.slice(0, indexOfDot);
            return {
              name: mainName,
              message
            };
          }
        });
      }

      return [];
    }

    throw error;
  }

  throw new Error("Couldn't process the request")
}