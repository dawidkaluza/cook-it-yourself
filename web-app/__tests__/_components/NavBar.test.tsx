import {beforeEach, describe, expect, test, vi} from 'vitest'
import {NavBar} from "@/app/_components/NavBar";
import {render, screen} from "@testing-library/react";
import {isSignedIn} from "@/app/_api/actions";
import {Mock} from "node:test";

vi.mock("@/app/_api/actions", async (importOriginal) => {
  const mod = await importOriginal<typeof import("@/app/_api/actions")>();

  return {
    ...mod,
    isSignedIn: vi.fn(),
  };
});

describe('NavBar component', () => {
  test('render when not signed in', () => {
    // Given
    (isSignedIn as Mock<any>).mockImplementation(() => false);

    // When
    render(<NavBar />);

    // Then
    const signInLink = screen.queryByRole("link", { name: "Sign in"});
    expect(signInLink).not.toBeNull();
    expect(signInLink?.getAttribute("href")).toBe("/sign-in")

    const navBarMenu = screen.queryByRole('list');
    expect(navBarMenu).toBeNull();
  })
});