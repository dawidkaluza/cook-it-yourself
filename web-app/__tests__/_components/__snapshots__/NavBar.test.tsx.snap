// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavBar component > render when signed in 1`] = `
<div>
  <nav
    class="navbar navbar-expand bg-body-tertiary"
  >
    <div
      class="container-fluid text-center"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        Cook it yourself
      </a>
      <button
        aria-controls="navbarNav"
        aria-expanded="false"
        aria-label="Toggle navigation"
        class="navbar-toggler"
        data-bs-target="#navbar-menu"
        data-bs-toggle="collapse"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="collapse navbar-collapse"
        id="navbar-menu"
      >
        <ul
          class="navbar-nav"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="/my-recipes"
            >
              My recipes
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</div>
`;

exports[`NavBar component > render when signed out 1`] = `
<div>
  <nav
    class="navbar navbar-expand bg-body-tertiary"
  >
    <div
      class="container-fluid text-center"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        Cook it yourself
      </a>
      <a
        class="btn btn-primary btn-sm"
        href="/sign-in"
      >
        Sign in
      </a>
    </div>
  </nav>
</div>
`;
